tests/test_main.py

```diff
@@ -2220,3 +2221,65 @@ class Model(BaseModel):
             also_protect_field: str
 
             model_config = ConfigDict(protected_namespaces=('protect_me_', 'also_protect_'))
+
+
+def parametrize_root_model():
+    class InnerModel(BaseModel):
+        int_field: int
+        str_field: str
+
+    return pytest.mark.parametrize(
+        ('root_type', 'root_value', 'dump_value'),
+        [
+            pytest.param(int, 42, 42, id='int'),
+            pytest.param(str, 'forty two', 'forty two', id='str'),
+            pytest.param(Dict[int, bool], {1: True, 2: False}, {1: True, 2: False}, id='dict[int, bool]'),
+            pytest.param(List[int], [4, 2, -1], [4, 2, -1], id='list[int]'),
+            pytest.param(
+                InnerModel,
+                InnerModel(int_field=42, str_field='forty two'),
+                {'int_field': 42, 'str_field': 'forty two'},
+                id='InnerModel',
+            ),
+        ],
+    )
+
+
+@parametrize_root_model()
+def test_root_model_specialized(root_type, root_value, dump_value):
+    Model = RootModel[root_type]
+
+    m = Model(root_value)
+
+    with pytest.warns(UserWarning, match='but got `dict` - serialized value may not be as expected'):
+        assert m.model_dump() == dump_value
```

Comment:
Model dump always complains about the root value data type stating that it should be `dict`. It looks like something isn't right on the `pydantic-core` side.

Reply:
I guess you'll need a custom serialised, but again it might be fixed with `root_model=True` in the core-schema.